apps/shared/avifpng.c

```diff
@@ -17,6 +18,106 @@ typedef png_bytep png_iccp_datap;
+static avifBool avifHexStringToBytes(const char * hexString, size_t expectedLength, avifRWData * bytes)
+{
+    avifRWDataRealloc(bytes, expectedLength);
+    const char * src = hexString;
+    uint8_t * dst = bytes->data;
+    size_t actualLength = 0;
+    for (; (actualLength < expectedLength) && (*src != '\0'); ++actualLength, ++dst) {
+        while (*src == '\n') {
+            ++src;
+        }
+        const char hexValue[3] = { src[0], src[1], '\0' };
+        src += 2;
+        char * end;
+        bytes->data[actualLength] = (uint8_t)strtol(hexValue, &end, 16);
+        if (end != &hexValue[2]) {
+            avifRWDataFree(bytes);
+            fprintf(stderr, "Exif extraction failed: invalid token at " AVIF_FMT_ZU "\n", actualLength);
+            return AVIF_FALSE;
+        }
+    }
+
+    if (actualLength != expectedLength) {
+        avifRWDataFree(bytes);
+        fprintf(stderr, "Exif extraction failed: expected " AVIF_FMT_ZU " tokens but got " AVIF_FMT_ZU "\n", expectedLength, actualLength);
+        return AVIF_FALSE;
+    }
+    return AVIF_TRUE;
+}
+
+// Parses the raw profile string of profileLen characters and extracts the payload.
+static avifBool avifCopyRawProfile(const char * profile, size_t profileLen, avifRWData * payload)
+{
+    if (!profile || (profileLen < 3)) {
+        fprintf(stderr, "Exif extraction failed: empty profile\n");
+        return AVIF_FALSE;
+    }
+
+    // ImageMagick formats 'raw profiles' as "\n<name>\n<length>(%8lu)\n<hex payload>\n".
+    const char * src = profile;
+    if (*src != '\n') {
+        fprintf(stderr, "Exif extraction failed: malformed raw profile, expected '\\n' but got '\\x%.2X'\n", *src);
+        return AVIF_FALSE;
+    }
+    ++src;
+    size_t numReadCharacters = 1;
+    // Skip the profile name and extract the length.
+    for (; (*src != '\0') && (*src != '\n'); ++src, ++numReadCharacters) {
+        if (numReadCharacters == profileLen) {
+            fprintf(stderr, "Exif extraction failed: truncated raw profile of size " AVIF_FMT_ZU "\n", profileLen);
+            return AVIF_FALSE;
+        }
+    }
+    char * end;
+    const long expectedLength = strtol(src, &end, 10);
+    if ((expectedLength <= 0) || ((unsigned long)expectedLength > SIZE_MAX)) {
+        fprintf(stderr, "Exif extraction failed: invalid length %ld\n", expectedLength);
+        return AVIF_FALSE;
+    }
+    if (*end != '\n') {
+        fprintf(stderr, "Exif extraction failed: malformed raw profile, expected '\\n' but got '\\x%.2X'\n", *end);
+        return AVIF_FALSE;
+    }
+    ++end;
+
+    // 'end' now points to the profile payload.
+    return avifHexStringToBytes(end, (size_t)expectedLength, payload);
+}
+
+// Returns AVIF_TRUE if there was no Exif metadata located at info or if the Exif metadata located at info
+// was correctly parsed and imported to avif->exif. Returns AVIF_FALSE in case of error.
+static avifBool avifExtractExif(png_structp png, png_infop const info, avifImage * avif)
```

Comment:
Does this support https://ftp-osl.osuosl.org/pub/libpng/documents/pngext-1.5.0.html#C.eXIf ?

Reply:
Doesn't look like it indeed.

There are macros to check if the libpng used is recent enough and has a method to support the eXIf chunk.

Or implement from scratch regardless of libpng used.

Reply:
Thanks for the links. I did not grasp everything on the eXIf topic. It is another kind of Exif chunk stored in a PNG, right? I assume libwebp does not handle those. I guess libavif should handle both Exif and eXIf chunks. Do you have an example image with an eXIf chunk?

I added a call to `png_get_eXIf_1()` but I did not find any code snippet using it on the web, so I have no idea how correct this is. "eXIf" being equivalent to "Exif" besides the case makes it hard to find documentation on the topic.

> Or implement own method from scratch regardless of libpng used. It is just a literal bytestream dump without any extra header nor coding like it is the case with this custom legacy scheme.

Could you precise what you meant:
- Implement a new way of parsing
  1. eXIf chunks in PNG files
  2. Exif or eXIf chunks in PNG files
- without relying on
  1. a version of libpng recent enough to support `png_get_eXIf_1()`.
  2. libpng at all.
- Any
  1. eXIf chunk
  2. Exif or eXIf chunk
- is just a literal bytestream dump without any extra header nor coding like it is the case with
  1. how it is handled in libwebp currently (`png_get_text()` + comparing `text->key` + raw profile parsing).
  2. how custom legacy Exif metadata chunks are laid out in PNG (raw profiles of the form `"\n<name>\n<length>(%8lu)\n<hex payload>\n"`).

Reply:
Yes, I meant in the case one builds libavif with a version of libpng that does not have support for it. OTOH, it might be valid to say "just use a more recent version of libpng" in order to have this chunk supported....

For eXIf chunk examples you can perhaps take a look at some Exiv2 test vectors.

Reply:
FWIW, Exiv2 will replace any eXIf chunk with a legacy raw one, while exiftool will always do the opposite, i.e. write the more modern eXIf chunk...